sycl/source/detail/scheduler/commands.cpp

```diff
@@ -156,58 +157,90 @@ getPiEvents(const std::vector<EventImplPtr> &EventImpls) {
-
-    // If contexts don't match - connect them using user event
-    if (DepEventContext != Context && !Context->is_host()) {
-      EventImplPtr GlueEvent(new detail::event_impl());
-      GlueEvent->setContextImpl(Context);
-      EventImplPtr *GlueEventCopy =
-          new EventImplPtr(GlueEvent); // To increase the reference count by 1.
-
-      RT::PiEvent &GlueEventHandle = GlueEvent->getHandleRef();
-      auto Plugin = Context->getPlugin();
-      auto DepPlugin = DepEventContext->getPlugin();
-      // Add an event on the current context that
-      // is triggered when the DepEvent is complete
-      Plugin.call<PiApiKind::piEventCreate>(Context->getHandleRef(),
-                                            &GlueEventHandle);
-
-      DepPlugin.call<PiApiKind::piEventSetCallback>(
-          DepEvent->getHandleRef(), PI_EVENT_COMPLETE, EventCompletionClbk,
-          /*void *data=*/(GlueEventCopy));
-      GlueEvents.push_back(GlueEvent);
-      Result.push_back(std::move(GlueEvent));
-      continue;
+
+    // wait for dependency host events
+    for (const EventImplPtr &Event : MThisCmd->MPreparedHostDepsEvents) {
+      Event->waitInternal();
     }
-    Result.push_back(DepEvent);
   }
-  MDepsEvents.insert(MDepsEvents.end(), GlueEvents.begin(), GlueEvents.end());
-  return Result;
+
+  // Lookup for empty command amongst users of this cmd
+  static EmptyCommand *findUserEmptyCommand(Command *ThisCmd) {
+    assert(ThisCmd->MUsers.size() == 1 &&
+           "Only a single user is expected for host task command");
+
+    Command *User = *ThisCmd->MUsers.begin();
+
+    assert(User->getType() == Command::CommandType::EMPTY_TASK &&
+           "Expected empty command as single user of host task command");
+    assert(User->MIsBlockable && "Empty command is expected to be blockable");
+    assert(User->MBlockReason == Command::BlockReason::HostTask &&
+           "Empty command is expected to be blocked due to host task");
+
+    return static_cast<EmptyCommand *>(User);
+  }
+
+public:
+  DispatchHostTask(ExecCGCommand *ThisCmd) : MThisCmd{ThisCmd} {}
+
+  void operator()() const {
+    waitForEvents();
+
+    assert(MThisCmd->getCG().get());
+    assert(MThisCmd->getCG()->getType() == CG::CGTYPE::CODEPLAY_HOST_TASK);
+
+    CGHostTask *HostTask = static_cast<CGHostTask *>(MThisCmd->getCG().get());
+
+    // we're ready to call the user-defined lambda now
+    HostTask->MHostTask->call();
+    HostTask->MHostTask.reset();
+
+    // unblock user empty command here
+    EmptyCommand *EmptyCmd = findUserEmptyCommand(MThisCmd);
+    assert(EmptyCmd && "No empty command found");
+
+    // Completing command's event along with unblocking enqueue readiness of
+    // empty command may lead to quick deallocation of MThisCmd by some cleanup
+    // process. Thus we'll copy deps prior to completing of event and unblocking
+    // of empty command.
+
+    std::vector<DepDesc> Deps = MThisCmd->MDeps;
```

Comment:
I think this triggers a segfault in some conditions, see my other comment below. 
Although you do copy dependencies into a local vector and free them afterwards, the contents of the Dependencies are not copied, and they may be released elsewhere. In the code I paste below, MRecord is null, because it has been released by another command.
I am not familiar enough with the entirety of the code but, why cannot the Leaves of the requirement be enqueued before marking this event as complete?

Reply:
Sorry, I realize mentioning below or above is confusing since in the pull request view it appears in the other way than in the file view!  :-)

Reply:
> why cannot the Leaves of the requirement be enqueued before marking this event as complete

This may lead to deadlock due to dependent commands waiting on `EmptyCmd`'s event.

Reply:
I see two distinct solutions here.
1. Store a copy of MRecord in `shared_ptr` to prevent it's deallocation.
2. Employ thread pool for enqueueing. I dunno if it'll help.

Reply:
This should be fixed now